pymc3/distributions/continuous.py

```diff
@@ -924,6 +925,67 @@ def logcdf(self, value):
         )
 
 
+class ZeroSumNormal(Continuous):
+    def __init__(self, sigma=1, zerosum_dims=None, zerosum_axes=None, **kwargs):
+        shape = kwargs.get("shape", ())
+        dims = kwargs.get("dims", None)
+        if isinstance(shape, int):
+            shape = (shape,)
+
+        if isinstance(dims, str):
+            dims = (dims,)
+
+        self.mu = self.median = self.mode = tt.zeros(shape)
+        self.sigma = tt.as_tensor_variable(sigma)
+
+        if zerosum_dims is None and zerosum_axes is None:
+            if shape:
+                zerosum_axes = (-1,)
+            else:
+                zerosum_axes = ()
+
+        if isinstance(zerosum_axes, int):
+            zerosum_axes = (zerosum_axes,)
+
+        if isinstance(zerosum_dims, str):
+            zerosum_dims = (zerosum_dims,)
+
+        if zerosum_axes is not None and zerosum_dims is not None:
+            raise ValueError("Only one of zerosum_axes and zerosum_dims can be specified.")
+
+        if zerosum_dims is not None:
+            if dims is None:
+                raise ValueError("zerosum_dims can only be used with the dims kwargs.")
+            zerosum_axes = []
+            for dim in zerosum_dims:
+                zerosum_axes.append(dims.index(dim))
+        self.zerosum_axes = [a if a >= 0 else len(shape) + a for a in zerosum_axes]
+
+        if "transform" not in kwargs or kwargs["transform"] == None:
+            kwargs["transform"] = transforms.ZeroSumTransform(zerosum_axes)
+
+        super().__init__(**kwargs)
+
+    def logp(self, value):
+        return Normal.dist(sigma=self.sigma).logp(value)
```

Comment:
If we don’t add the scaling of sigma here, our `random` method will be inconsistent with the logp

Reply:
How so ? Isn't `random` also drawing samples using `self.sigma` ?

Reply:
This logp is somewhat strange still. From the math side this should be pm.MvNormal with cov=I - J / n where J is a matrix of all 1s. We don't want to write it like this though, because we don't want to do matrix factorization, and pm.MvNormal doesn't work if an eigenvalues is 0.
It would be great, if instead we could define the logp simply in the transformed space. This would imply changes to TransformedDistribution though.

Reply:
We can avoid the `MvNormal.logp` problem if we force `self.sigma` to be a scalar or to have a single element in all the `zerosum_axes`. In this case, all the directions in the zerosum manifold are uncorrelated and have equal variance. This means that we can use the `Normal.logp` as long as we also include a `bound` condition that guarantees that we are on the zerosum manifold. You can do this by using this logp:

```suggestion
        zerosums = [tt.all(tt.abs_(tt.mean(x, axis=axis)) <= 1e-9) for axis in self.zerosum_axes]
        return bound(
            pm.Normal.dist(sigma=self.sigma * self._rescaling).logp(x),
            tt.all(self.sigma > 0),
            broadcast_conditions=False,
            *zerosums,
        )
```

Reply:
I came across [this wiki section](https://en.wikipedia.org/wiki/Multivariate_normal_distribution#Degenerate_case) that talks about the degenerate MvNormal case (which is what we have with the ZeroSumNormal). We could use that formula as the expected logp value and test if the `logp` that we are using in the distribution matches it. The expected logp would look something like this:

```python
def pseudo_log_det(A, tol=1e-15):
    w, v = np.linalg.eigh(A)
    return np.sum(np.log(np.where(np.abs(v) >= tol, v, 1)), axis=-1)

def logp(value, sigma):
    n = value.shape[-1]
    cov = sigma[..., None, None]**2 * (np.eye(n) - np.ones((n, n)) / n)
    x = value[..., None]
    return 0.5 * pseudo_log_det(2 * np.pi * cov) - 0.5 * np.move_axis(x, -2, -1) @ np.pinv(cov) @ x
```

Reply:
I ran a few tests with the logp and it looks like the `logp` that we are using in this PR, doesn't match what one would expect from a degenerate multivariate normal distribution. In my comment above, I posted what a degenerate MvNormal `logp` looks like. For this particular problem, where we know that we have only one eigenvector with zero eigenvalue, we can re-write the logp as:

```python
def logp(value, sigma):
    n = value.shape[-1]
    cov = np.asarray(sigma)[..., None, None]**2 * (np.eye(n) - np.ones((n, n)) / n)
    v, w = np.linalg.eigh(cov)
    psdet =  0.5 * (np.sum(np.log(v[..., 1:])) + (n - 1) * np.log(2 * np.pi))
    cov_pinv = w[:, 1:] @ np.diag(1 / v[1:]) @ w[:, 1:].T
    exp = 0.5 * (value[..., None, :] @ cov_pinv @ value[..., None])[..., 0, 0]
    return np.where(np.abs(np.sum(value, axis=-1)) < 1e-9, -psdet - exp, -np.inf)
```

This is different from the `logp` that we are currently using in this PR. The difference is in the normalization constant:
`psdet =  0.5 * (np.sum(np.log(v[..., 1:])) + (n - 1) * np.log(2 * np.pi))`. In particular, since, all eigenvalues `v` except the first one are the same and are equal to `sigma**2`, `psdet = (n - 1) * (0.5 * np.log(2 * np.pi) + np.log(np.sigma))`. Whereas, with the assumed `pm.Normal.dist(sigma=self.sigma).logp(x)` the normalization factor we are getting is:

```python
psdet = n * (0.5 * np.log(2 * np.pi) + np.log(np.sigma))
```

This means that we have to multiply the `logp` that we are using by (n-1)/n (in the case where only one axis sums to zero) to get the correct log probability density.